### ISSUE TITLE
Qwen-Edit-2509 : "Issue: Dimension mismatch in LanPaint KSampler"

### ISSUE THREAD

**Issue by olPlow** [scraed/LanPaint#64]
Hi 

<img width="1559" height="441" alt="Image" src="https://github.com/user-attachments/assets/ab8af580-670a-4bfe-8e37-4c890caf86cb" />

!!

I am encountering a spatial dimension error when using a mask with 'Set Latent Noise'. The process works correctly with an 'Empty Latent', suggesting the mask input is causing the conflict.

Error Message:

Input and output must have the same number of spatial dimensions, but got input with spatial dimensions of [1, 175, 93] and output size of torch.Size([175, 93]).

It looks like the input tensor format (N, C, H, W) is not aligning with the expected output size. Any advice on how to fix this? (Setup screenshot attached)."

Thanks !!

**Comment by scraed** [comment]
For qwen image, the input image should have shape (N, C, 1, H, W), the 1 here is for time dimension as it share similar architecture with Wan. Your resized image as well as mask should have that extra 1 dimension. I suggest you convert mask to image, resize, then convert back to mask, like the example I provided in templates.

**Comment by olPlow** [comment]
Thanks for getting back to me so quickly. Unfortunately, I'm still getting the exact same error even when using the 'Masked_Qwen_Image_Edit_2509' template, Please find the error log below for reference : 

# ComfyUI Error Report
## Error Details
- **Node ID:** 125
- **Node Type:** LanPaint_KSampler
- **Exception Type:** ValueError
- **Exception Message:** Input and output must have the same number of spatial dimensions, but got input with spatial dimensions of [1, 175, 93] and output size of torch.Size([175, 93]). Please provide input tensor in (N, C, d1, d2, ...,dK) format and output size in (o1, o2, ...,oK) format.

## Stack Trace
```
  File "/workspace/ComfyUI/execution.py", line 516, in execute
    output_data, output_ui, has_subgraph, has_pending_tasks = await get_output_data(prompt_id, unique_id, obj, input_data_all, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb, v3_data=v3_data)
                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/workspace/ComfyUI/execution.py", line 330, in get_output_data
    return_values = await _async_map_node_over_list(prompt_id, unique_id, obj, input_data_all, obj.FUNCTION, allow_interrupt=True, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb, v3_data=v3_data)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/workspace/ComfyUI/custom_nodes/ComfyUI-Lora-Manager/py/metadata_collector/metadata_hook.py", line 165, in async_map_node_over_list_with_metadata
    results = await original_map_node_over_list(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/workspace/ComfyUI/execution.py", line 304, in _async_map_node_over_list
    await process_inputs(input_dict, i)

  File "/workspace/ComfyUI/execution.py", line 292, in process_inputs
    result = f(**inputs)
             ^^^^^^^^^^^

  File "/workspace/ComfyUI/custom_nodes/LanPaint/src/LanPaint/nodes.py", line 324, in sample
    return nodes.common_ksampler(model, seed, steps, cfg, sampler_name, scheduler, positive, negative, latent_image, denoise=denoise)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/workspace/ComfyUI/nodes.py", line 1505, in common_ksampler
    samples = comfy.sample.sample(model, noise, steps, cfg, sampler_name, scheduler, positive, negative, latent_image,
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/workspace/ComfyUI/comfy/sample.py", line 60, in sample
    samples = sampler.sample(noise, positive, negative, cfg=cfg, latent_image=latent_image, start_step=start_step, last_step=last_step, force_full_denoise=force_full_denoise, denoise_mask=noise_mask, sigmas=sigmas, callback=callback, disable_pbar=disable_pbar, seed=seed)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/workspace/ComfyUI/comfy/samplers.py", line 1178, in sample
    return sample(self.model, noise, positive, negative, cfg, self.device, sampler, sigmas, self.model_options, latent_image=latent_image, denoise_mask=denoise_mask, callback=callback, disable_pbar=disable_pbar, seed=seed)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/workspace/ComfyUI/comfy/samplers.py", line 1068, in sample
    return cfg_guider.sample(noise, latent_image, sampler, sigmas, denoise_mask, callback, disable_pbar, seed)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/workspace/ComfyUI/comfy/samplers.py", line 1050, in sample
    output = executor.execute(noise, latent_image, sampler, sigmas, denoise_mask, callback, disable_pbar, seed, latent_shapes=latent_shapes)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


## System Information
- **ComfyUI Version:** 0.5.1
- **Arguments:** main.py --listen 0.0.0.0 --port 8188
- **OS:** linux
- **Python Version:** 3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]
- **Embedded Python:** false
- **PyTorch Version:** 2.8.0+cu128
  File "/workspace/ComfyUI/comfy/patcher_extension.py", line 112, in execute
    return self.original(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/workspace/ComfyUI/custom_nodes/LanPaint/src/LanPaint/nodes.py", line 94, in outer_sample
    denoise_mask = prepare_mask(denoise_mask, noise.shape, device, video_inpainting)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/workspace/ComfyUI/custom_nodes/LanPaint/src/LanPaint/nodes.py", line 64, in prepare_mask
    return reshape_mask(noise_mask, shape,video_inpainting).to(device)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/workspace/ComfyUI/custom_nodes/LanPaint/src/LanPaint/nodes.py", line 56, in reshape_mask
    mask = torch.nn.functional.interpolate(input_mask, size=output_shape[-2:], mode=scale_mode)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/dist-packages/torch/nn/functional.py", line 4647, in interpolate
    raise ValueError(

```

**Comment by aerocat135** [comment]
I'm also getting the same error.  I had not used the LanPaint KSampler workflow I have for QWEN 2509 in a week or so, but when I first ran it, it worked, and quite well.  I don't know exactly what happened between now and then - I don't think anything changed for the KSampler node.  What I think -might- have changed is the code for the `Set Latent Noise Mask` node.  I had bypassed that node and the error didn't happen and the KSampler started to process.  However, since I think that was going to try to have it inpaint the entire image, I cancelled the run.  Anyway, the way the error shows up, it is showing the output tensor shape as having one less dimension than the input tensor shape that is coming from the `Set Latent Noise Mask` node, i.e., the output tensor for the KSampler is of the form [H,W] whereas the input tensor is of the form [B, H, W].

Let me start this next part by saying I am really not a Python coder, nor am I particularly knowledgeable about the custom node coding.  But through a little bumbling around on Google, it appears that there in Line 56 of `ComfyUI/custom_nodes/LanPaint/src/LanPaint/nodes.py` , the argument `size=output_shape[-2:]` is culling out just the last two dimensions [H,W] for the output tensor shape.  As an experiment, I changed that to `size=output_shape[-3:]`  (DEFINITELY saved a backup copy of that nodes.py) so that the output tensor now has three dimensions to at least match the input tensor.  I can verify that the error did not occur and the KSampler is currently processing.  I can't say for sure if it is doing the inpainting correctly, as my machine is woefully slow and is still processing.  I don't know if adding to the output tensor dimensions is the actual solution or if somehow the input tensor needs to be shortened.  It may be that by increasing the output tensor dimensions I'm making it use incorrect values for the inpaint process.  We'll see.  Again...I don't think the actual error is because of anything that happened to the LanPaint KSampler, it looks like something changed in the `Set Latent Noise Mask` node that appears to be part of the comfyui-core, and likely got an update in a recent update to ComfyUI.

**Comment by CPapadim** [comment]
Same issue.

@aerocat135 your general diagnosis is correct though the specific code is just one node over I believe.

The code on line 56 wasn't changed since early October and the Set Latent Noise Mask node outputs don't seem recently changed either.  The VAE encode dimensions do seem to have changed though on December 18th.  I installed the December 15th commit (33c7f1179d4a961e4ca1dd78188c5134e0ee8e8c) of ComfyUI and that seems to work so I think with high certainty those VAE node changes are the culprit

**Comment by scraed** [comment]
Which comfyui version you guys are using? I think that is important. Let me try to reproduce this error.

**Comment by CPapadim** [comment]
> Which comfyui version you guys are using? I think that is important. Let me try to reproduce this error.

I used the latest (as of today) and had the issue.

I used the version on the commit I shared here:  33c7f1179d4a961e4ca1dd78188c5134e0ee8e8c  before the VAE changes made in the following commit, and did not have the issue.  

I did not check any in between commits, but I'm betting the VAE encode changes in nodes.py in comfyui on this commit (https://github.com/comfyanonymous/ComfyUI/commit/6a2678ac65ff690e24771a4c64ce96f7a9824fa4#diff-e485eb8bfa77e53e217e9b400c67298309ef679c6ecd69a1174366f48bb93499) are the issue.

**Comment by aerocat135** [comment]
@scraed,

It looks like my ComfyUI version is 0.5.1.  And it could very well be as CPapadim says that it is something about the VAE Encode node for the mask latent that is causing the issue.  The December 18th timeframe pretty much fits with when things were working and then afterward were not, for me.  My reasoning though is that when using the original `size=output_shape[-2:]` bypassing just the `Set Latent Noise Mask` node alone is enough to keep the error from popping up and the KSampler will go ahead and start processing the unmasked latent coming out of the VAE Encode.  My test run with setting `size=output_shape[-3:]` with the `Set Latent Noise Mask` node active did process complete and seemed to inpaint the masked area only.  I'm still not sure that my tweak to add the "B" value to the output tensor is the correct thing to do.  It may be that removing the "B" index from the input tensor [B, H, W] is more correct, since the inpainting using just the height and width [H, W] worked just fine before whatever changed had changed.  I don't know if one solution is more 'future proof' against any new changes that might occur in ComfyUI.  And in case it gets lost in all this discussion, I do want to say that I appreciate your work on this - the LanPainting does work very well.  Just seems to have been scuttled temporarily :).  I do hope that my tweaking around and commenting are more help than harm.

**Comment by olPlow** [comment]
@aerocat135 

Thank you so much for this insight! I tested your fix by changing Line 56 in nodes.py to size=output_shape[-3:] as you suggested, and I can confirm that it works perfectly. The error is gone, and more importantly, the inpainting is processing correctly on my end. 

It’s working fine with the latest ComfyUI version for now. Hopefully, this fix remains compatible with future updates. :)

**Comment by aerocat135** [comment]
Well, you're welcome, though I found your post while looking for a solution to the same thing, myself.  With all due respect to scraed, though, it is more of an inelegant workaround than a solution.  It does seem to work without breaking anything else, but I don't know for sure whether or not it impacts other calculations within the KSampler processes.  This section of code is for the condition of a 2D image (in the 'else' clause of a conditional statement).  The section of code above this line reshapes the input tensor for video (I think) so the more elegant solution might be to reshape the 2D image input tensor and let LanPaint work as originally intended.  I'm actually sort of talking above my own head, and I hope scraed does not take offense at my hacking around on the code.

**Comment by CPapadim** [comment]
> Well, you're welcome, though I found your post while looking for a solution to the same thing, myself. With all due respect to scraed, though, it is more of an inelegant workaround than a solution. It does seem to work without breaking anything else, but I don't know for sure whether or not it impacts other calculations within the KSampler processes. This section of code is for the condition of a 2D image (in the 'else' clause of a conditional statement). The section of code above this line reshapes the input tensor for video (I think) so the more elegant solution might be to reshape the 2D image input tensor and let LanPaint work as originally intended. I'm actually sort of talking above my own head, and I hope scraed does not take offense at my hacking around on the code.

You should give yourself more credit, because of your research and workaround the developer will have a much easier time creating a solution.  Even if your change ends up not being the one that sticks, it identified the cause of what could have been a hard to reproduce issue.  And in the meantime people can use your workaround or an earlier version of comfy before the shape changes to do their generations.  Great sleuthing work!

**Comment by zslefour** [comment]
> Which comfyui version you guys are using? I think that is important. Let me try to reproduce this error.

It was fine yesterday, but after upgrading ComfyUI to the latest version v0.6.0 today, the same error occurred.

**Comment by zslefour** [comment]
> > Which comfyui version you guys are using? I think that is important. Let me try to reproduce this error.
> 
> It was fine yesterday, but after upgrading ComfyUI to the latest version v0.6.0 today, the same error occurred.

Rolling back to version v0.5.1 works fine

**Comment by Azrael234** [comment]
v0.6.0 also encountered this issue, hope it can be resolved.

**Comment by scraed** [comment]
@aerocat135 Nice fix. It indeed is the reason.
I have been avoiding changing comfy's code directly. Figured out a way to fixed it inside LanPaint node. Please try the 1.4.8

**Comment by aerocat135** [comment]
I'm happy to have helped in any way.  I had only just started playing with QWEN and was happy to have had your node.  I guess I have sort of gotten used to digging into the code when things go south on my machine; I'm actually running an older AMD card with Comfy-Zluda, so it is not uncommon for me to have to patch things occasionally.  Just getting ComfyUI to run on my machine the first place was sort of its own adventure and only accomplished with the help of folks in the GitHub community.

Thank you again....I'll update to 1.4.8 on my next re-boot of things :)
